[MOB-22537] - added processing of listeners asynchronously inside serial queue

akhiljain1907 opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
This PR solves race condition issue observed in Optimize SDK.
Jira Link - [MOB-22537](https://jira.corp.adobe.com/browse/MOB-22537)

When a personalization request is made to edge, our propositionsInProgress dictionary is returned from parent's repsonseEvent  before it gets updated in processEdgeResponse. We are now processing our listeners in a serial queue to solve this. 

<!--- Describe your changes in detail -->

## Related Issue

<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->

## Motivation and Context

<!--- Why is this change required? What problem does it solve? -->

## How Has This Been Tested?

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Screenshots (if appropriate):

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

sbenedicadb commented on Sources/AEPOptimize/Optimize.swift:
```diff
@@ -205,112 +198,118 @@ public class Optimize: NSObject, Extension {
-                      "Cannot process the update propositions request event, Configuration shared state is not available.")
-            return
-        }
+        queue.async {
```
should we capture `[weak self]` here as well to prevent a strong reference to `self` on line 203?

akhiljain1907 replied:
Done

sbenedicadb commented on Podfile.lock:
```diff
@@ -2,9 +2,9 @@ PODS:
     - AEPCore (< 6.0.0, >= 5.0.0)
     - AEPServices (< 6.0.0, >= 5.0.0)
-  - AEPCore (5.3.1):
+  - AEPCore (5.4.0):
```
let's be sure these updated dependency versions are represented in the podspec and Package.swift files

akhiljain1907 replied:
Done

sbenedicadb commented on Tests/AEPOptimizeTests/FunctionalTests/OptimizeFunctionalTests.swift:
```diff
@@ -181,9 +181,13 @@ class OptimizeFunctionalTests: XCTestCase {
-        XCTAssertEqual(1, updateEventIdsInProgress.count)
-        XCTAssertEqual([DecisionScope(name: "eyJhY3Rpdml0eUlkIjoieGNvcmU6b2ZmZXItYWN0aXZpdHk6MTExMTExMTExMTExMTExMSIsInBsYWNlbWVudElkIjoieGNvcmU6b2ZmZXItcGxhY2VtZW50OjExMTExMTExMTExMTExMTEifQ==")], updateEventIdsInProgress.values.first)
+        // using DispatchQueue to change the run loop as the events are now being processed inside a serial queue in optimize extension
+        DispatchQueue.main.async { [weak self] in
```
probably better to run this on a global (bg) thread than the main thread

akhiljain1907 replied:
I tried using background thread, but on running the tests there were inconsistency as a few of them used to fail because of the same reason we used a DispatchQueue to change the run loop (Assertions are executed before the events get processed).

sbenedicadb reviewed (changes requested):
couple small comments that need to be addressed

akhiljain1907 commented:
Hi @sbenedicadb thanks a lot for reviewing. I have addressed your comments and made relevant changes in the PR. Would request you to kindly check and review again.
